src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java

```diff
@@ -0,0 +1,180 @@
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.server.permission.nodes;
+
+import com.google.common.base.Preconditions;
+import net.minecraft.network.chat.Component;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraftforge.server.permission.events.PermissionGatherEvent;
+import net.minecraftforge.server.permission.handler.IPermissionHandler;
+
+import javax.annotation.Nonnull;
+import java.util.Objects;
+
+/**
+ * Represents the basic unit at the heart of the permission system.
+ *
+ * <p>A permission indicates the ability for an actor to perform an action,
+ * in its most general sense. In the permission system, all permissions are encoded as instances of this class,
+ * optionally integrated by a {@link PermissionDynamicContext}.</p>
+ *
+ * <p>A node is uniquely identified by its `nodeName`,
+ * which is a dot-separated string providing meaning to the node itself.
+ * The suggested structure of the name is `modid.name`, where `modid` is the ID of the owner of the node.
+ * This API <strong>does not require</strong> any implicit hierarchy,
+ * so `modid.name` is not considered a parent of `modid.name.sub`. Such decisions are left to the {@link IPermissionHandler}.</p>
+ *
+ * <p>Each node also has an associated {@link PermissionType}, detailing its exact type,
+ * along with a {@link PermissionResolver} that is used to obtain the default value of the permission.
+ * More information can be found on their documentation.</p>
+ *
+ * <p>A node can also be bound to additional {@link PermissionDynamicContextKey}s,
+ * which are used when querying the availability of the permission.
+ * There are no restrictions on their amount. It is also not mandatory to provide a value for all dynamics in a permission query.
+ * See the above link for more information.</p>
+ *
+ * <p>Each node should be registered via the {@link PermissionGatherEvent.Nodes} and stored statically in a field.
+ * That instance should then be reused every-time a permission check needs to be performed via {@link net.minecraftforge.server.permission.PermissionAPI#getPermissions(ServerPlayer, PermissionNode, PermissionDynamicContext[])}.</p>
+ */
+public final class PermissionNode<T>
+{
+    private final String nodeName;
+    private final PermissionType<T> type;
+    private final PermissionResolver<T> defaultResolver;
+    private final PermissionDynamicContextKey<?>[] dynamics;
+
+    private Component readableName;
+    private Component description;
```

Comment:
As these are `null` by default, these should be marked `@Nullable` along with their getters.